main.go

```diff
@@ -1,7 +1,352 @@
+	var user User
+	if err := json.NewDecoder(r.Body).Decode(&user); err != nil {
+		panic(err)
+	}
+	//Todo (Farouk): Mock ID - not safe
+	user.ID = strconv.Itoa(rand.Intn(1000000))
+	users = append(users, user)
+}
+
+// createTransfer route handler create transfer PATCH method
+func createTransfer(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	for _, item := range certificates {
+		if item.ID == params["id"] {
+			certificate := &item
+			var c Certificate
+			if err := json.NewDecoder(r.Body).Decode(&c); err != nil {
+				fmt.Fprintf(w, err.Error())
+				return
+			}
+			(*certificate).Transfer.To = c.Transfer.To
+			(*certificate).Transfer.Status = c.Transfer.Status
+			json.NewEncoder(w).Encode(*certificate)
+		}
+		return
+	}
+}
+
+// acceptTransfer route handler accept transfer PATCH method
+func acceptTransfer(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	params := mux.Vars(r)
+	userEmail, _, _ := r.BasicAuth()
+	for index, item := range certificates {
+		fmt.Println("Checking item " + item.ID + "againgst params ID: " + params["id"])
+		if item.ID == params["id"] && item.Transfer.To == userEmail {
+			fmt.Println("Certificate ID and userEmail match.")
+			certificate := &item
+			fmt.Println("Certificate.ID: " + (*certificate).ID)
+			var err error
+			coid, err := lookUpUserIDByEmail(userEmail)
+			if err != nil {
+				fmt.Fprintf(w, err.Error())
+				return
+			}
+			(*certificate).OwnerID = coid
+			(*certificate).Transfer.To = ""
+			(*certificate).Transfer.Status = "accepted transfer"
+			certificates = append(certificates[:index], certificates[index+1:]...)
+			certificates = append(certificates, (*certificate))
+			if err := json.NewEncoder(w).Encode(*certificate); err != nil {
+				fmt.Fprintf(w, err.Error())
+				return
+			}
+			break
+		}
+	}
+
+}
+
+// basicAuth basic http auth against email and password
+func basicAuth(handler http.HandlerFunc, userEmail string, password string, realm string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		u, pw, ok := r.BasicAuth()
+		if !ok || subtle.ConstantTimeCompare([]byte(u), []byte(userEmail)) != 1 ||
+			subtle.ConstantTimeCompare([]byte(pw), []byte(password)) != 1 {
+			w.Header().Set("WWW-Authenticate", `Basic realm="`+realm+`"`)
+			w.WriteHeader(401)
+			w.Write([]byte("Unauthorised.\n"))
+			return
+		}
+
+		handler(w, r)
+	}
+}
+
+// Main
 func main() {
-	fmt.Println("Hello World")
+	fmt.Println("Running...")
+
+	initInMemoryData()
+
+	r := mux.NewRouter()
+	userA, userB := 0, 1
+	r.HandleFunc("/", homeHandler).Methods("GET")
+	r.HandleFunc("/certificates", getCertificates).Methods("GET")
+	r.HandleFunc("/users/{id}/certificate", createCertificate).Methods("POST")
+	r.HandleFunc("/certificate/{id}", getCertificate).Methods("GET")
+	r.HandleFunc("/users/{userId}/certificate/{id}", patchCertificate).Methods("PATCH")
```

Comment:
POST, PUT and PATCH all have json data to decode. This `acceptTransfer` end point handler does not have any JSON to decode. The only information needed is in the url. So this is a GET.

Why? we are json.Encoding the certificate and it's being returned. We are "GET"ing that item, whilst more importantly triggering ownership state transfer.